wagtail/contrib/search_promotions/views.py

```diff
@@ -2,85 +2,104 @@
 
-@any_permission_required(
-    "wagtailsearchpromotions.add_searchpromotion",
-    "wagtailsearchpromotions.change_searchpromotion",
-    "wagtailsearchpromotions.delete_searchpromotion",
-)
-@vary_on_headers("X-Requested-With")
-def index(request):
-    # Ordering
-    valid_ordering = ["query_string", "-query_string", "views", "-views"]
-    ordering = valid_ordering[0]
-
-    if "ordering" in request.GET and request.GET["ordering"] in valid_ordering:
-        ordering = request.GET["ordering"]
-
-    # Query
-    queries = Query.objects.filter(editors_picks__isnull=False).distinct()
-
-    if "views" in ordering:
-        queries = queries.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
 
-    queries = queries.order_by(ordering)
-
-    # Search
+@method_decorator(
+    any_permission_required(
+        "wagtailsearchpromotions.add_searchpromotion",
+        "wagtailsearchpromotions.change_searchpromotion",
+        "wagtailsearchpromotions.delete_searchpromotion",
+    ),
+    name="dispatch",
+)
+@method_decorator(vary_on_headers("X-Requested-With"), name="dispatch")
+class IndexView(WagtailIndexView):
+    template_name = "wagtailsearchpromotions/index.html"
+    result_template_name = "wagtailsearchpromotions/results.html"
+    paginate_by = 20
     is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                    placeholder=_("Search promoted results"),
-                ),
-            },
-        )
+    ordering = "query_string"
+
+    def get_ordering(self):
+        # Ordering
+        valid_ordering = ["query_string", "-query_string", "views", "-views"]
+        ordering = self.ordering
+
+        if (
+            "ordering" in self.request.GET
+            and self.request.GET["ordering"] in valid_ordering
+        ):
+            ordering = self.request.GET["ordering"]
+        return ordering
```

Comment:
```suggestion
    def get_valid_orderings():
        return ["query_string", "-query_string", "views", "-views"]
```

You should not have to do any of this work, the Django Listing `MultipleObjectMixin` does most of this for  us, we should just have to se the valid orderings only.